### PR TITLE
[REVIEW] CSV Reader: Fix days being off-by-one when parsing some dates

### LINKED ISSUE #546: [BUG] Parsing dates with read_csv adds 1 day to dates
Parsing dates with read_csv adds 1 day to date.

```python
>>> with open('test.csv', 'w') as f:
...   f.write('1970-1-1,1970-1-2\n')
... 
18
>>> from cudf import read_csv
>>> df = read_csv('test.csv', names=list('ab'), dtype=['date','date']) 
>>> print(df)
                        a                       b
0 1970-01-02T00:00:00.000 1970-01-03T00:00:00.000
>>> df.a.astype('q')[0] == 24 * 3600 * 1000
True
```

```python
>>> # conversion from Python's date works:
>>> from datetime import date
>>> from cudf import DataFrame as DF
>>> DF({'a': [date(1970,1,1)]}).a.astype('q')[0] == 0
True
```

**Environment details (please complete the following information):**
 - Docker
 - Ubuntu 16.04 amd64
 - P100 and driver 396.44
 - CUDA: 9.2
 - Method of cuDF install: from source (commit 7aa3799e73948899a85e70bc73d3cea50b65dff1)
 - gcc/g++ (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
 - cmake version 3.13.1

### PR DESCRIPTION
**Pull request by j-ieong**
Fixes #546.

- Change daysSinceEpoch() algorithm to the one referenced in datetime_ops.cu
  - formula itself is pretty nifty, plus it seems good to have the same source for date/timestamp calculations as the rest of cuDF
- Add the problematic dates to the existing dates pytest
- Fix 'firstOcurance()' spelling error and move it to datetime_parser.cuh

### CODE REVIEW

File: cpp/src/io/csv/datetime_parser.cuh [hunk 1]
```diff
@@ -321,43 +345,32 @@ bool extractTime(char *data, int sIdx, int eIdx, int *hour, int *minute, int *se
...
-	// days up to start of month
-	days_e += days[me];
-
-	// now just add days, but not current full days since this one is not over
-	days_e +=  day;
-
-	return days_e;
+constexpr int32_t daysSinceEpoch(int year, int month, int day) {
+
+    // More details of this formula is located in cuDF datetime_ops
```
Comment (harrism) [inline cpp/src/io/csv/datetime_parser.cuh]:
```suggestion
    // More details of this formula are located in cuDF datetime_ops
```

File: cpp/src/io/csv/datetime_parser.cuh [hunk 2]
```diff
@@ -321,43 +345,32 @@ bool extractTime(char *data, int sIdx, int eIdx, int *hour, int *minute, int *se
...
+    // The months are shifted so that Feb (possible leap day) is the last month
+    // The result subtraction of 719468 shifts start date to epoch (1970-01-01)
+    year -= (month <= 2) ? 1 : 0;
+
+    const int32_t era = (year >= 0 ? year : year - 399) / 400;
+    const int32_t yoe = year - era * 400;
+    const int32_t doy = (153 * (month + (month > 2 ? -3 : 9)) + 2)/5 + day - 1;
+    const int32_t doe = yoe * 365 + yoe/4 - yoe/100 + doy;
+
+    return era * 146097 + doe - 719468;
```
Comment (harrism) [inline cpp/src/io/csv/datetime_parser.cuh]: Lots of magic numbers in this code. Do you think they could be named and replaced by constexpr variables?
Seems like a perfect use for C++1 user-defined literals. https://en.cppreference.com/w/cpp/language/user_literal

**Comment by vuule** [comment]
looks good

File: cpp/src/io/csv/datetime_parser.cuh (hunk 2 above)
Comment (j-ieong) [reply cpp/src/io/csv/datetime_parser.cuh]: That's a cool idea. Let me work on that, if this fix isn't urgent. There's lot of magic numbers in datetime_ops too.

File: cpp/src/io/csv/datetime_parser.cuh [hunk 3]
```diff
@@ -306,58 +330,88 @@ bool extractTime(char *data, int sIdx, int eIdx, int *hour, int *minute, int *se
...
+    // In brief, the calculation is split over several components:
+    //     era: a 400 year range, where the date cycle repeats exactly
+    //     yoe: year within the 400 range of an era
+    //     doy: day within the 364 range of a year
+    //     doe: exact day within the whole era
+    // The months are shifted so that March is the starting month and February
+    // (possible leap day in it) is the last month for the linear calculation
+    year -= (month <= 2) ? 1 : 0;
+
+    const int32_t era = (year >= 0 ? year : year - 399_years) / 1_erasInYears;
```
Comment (harrism) [inline cpp/src/io/csv/datetime_parser.cuh]: Before I merge, I want to make sure you are definitely intending integer division throughout this computation?

Comment (j-ieong) [reply cpp/src/io/csv/datetime_parser.cuh]: Yes, the original whitepaper intends it to be integer division. We could simplify it if we didn't want to handle negative years but I wanted to keep the algorithm mostly intact.

Comment (harrism) [reply cpp/src/io/csv/datetime_parser.cuh]: OK!

**Pull request merged**